### PR TITLE
Post: nanoparquet 0 3 0

### PR DESCRIPTION
**Pull request by gaborcsardi**

### CODE REVIEW

File: content/blog/nanoparquet-0-3-0/index.Rmd [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+* **Well supported**. At this point Parquet is well supported across
+  modern languages like R, Python, Rust, Java, Go, etc. In particular,
+  Apache Arrow handles Parquet files very well, and has bindings to many
+  languages. DuckDB is a very portable tool that reads and writes Parquet
+  files, or even opens a set of Parquet files as a database.
+
+* **Performant**. Parquet columns may use various encodings and compression
+  to ensure that the data files are kept as small as possible. When running
+  an analytical query on the subset of the data, the Parquet format makes
+  it easy to skip the columns and/or rows that are irrelevant.
```
Comment (hadley) [inline content/blog/nanoparquet-0-3-0/index.Rmd:86]: Mention something about parallelism too?

File: content/blog/nanoparquet-0-3-0/index.Rmd [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+and is easy to install. Our goal is to facilitate the adoption of Parquet
+for smaller data sets, especially for teams that share data between
+multiple environments, e.g. R, Python, Java, etc.
+
+## nanoparquet Features
+
+These are some of the nanoparquet features that we are most excited about.
+
+* **Lightweight**. nanoparquet has no package or system dependencies other
+  than a C++-11 compiler. It compiles in about 30 seconds, into an R
```
Comment (hadley) [inline content/blog/nanoparquet-0-3-0/index.Rmd:110]:
```suggestion
  than a C++-11 compiler. It compiles in about 30 seconds into an R
```

File: content/blog/nanoparquet-0-3-0/index.Rmd [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+  package that is less than a megabyte in size.
+
+* **Reads many Parquet files**. `nanoparquet::read_parquet()` supports
+  reading most Parquet files. In particular, in supports all Parquet
+  encodings and at the time of writing it supports three compression
+  codecs: Snappy, Gzip and Zstd. Make sure you read "Limitations" below.
+
+* **Writes many R data types**. `nanoparquet::write_parquet()` supports
+  writing most R data frames. In particular, missing values are handled
+  properly, factor columns are keps as factors, temporal types as well.
```
Comment (hadley) [inline content/blog/nanoparquet-0-3-0/index.Rmd:120]:
```suggestion
  properly, factor columns are keps as factors, and temporal types are encoded correctly.
```

Or similar

File: content/blog/nanoparquet-0-3-0/index.Rmd [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+
+* **Type mappings**. nanoparquet has a well defined set of
+  [type mapping rules](https://r-lib.github.io/nanoparquet/reference/nanoparquet-types.html).
+  Use the [`parquet_column_types()`](https://r-lib.github.io/nanoparquet/dev/reference/parquet_column_types.html)
+  function to see how `read_parquet()` and `write_parquet()` maps
+  Parquet and R types for a file or a data frame.
+
+* **Metadata queries**. nanoparquet has a
+  [number of functions](https://r-lib.github.io/nanoparquet/dev/reference/index.html#extract-parquet-metadata)
+  to query the metadata and schema in Parquet files.
```
Comment (hadley) [inline content/blog/nanoparquet-0-3-0/index.Rmd:131]:
```suggestion
* **Metadata queries**. nanoparquet has a
  [number of functions](https://r-lib.github.io/nanoparquet/dev/reference/index.html#extract-parquet-metadata)
  that allow you to query the metadata and schema without reading in the full dataset.
```

File: content/blog/nanoparquet-0-3-0/index.Rmd [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+will create for that column. If `repetition_type` is `REQUIRED`, then that
+column cannot contain missing values. `OPTIONAL` columns may have missing
+values. `logical_type` is the higher level Parquet data type.
+
+E.g. the first column is an UTC (because of the `TRUE`) timestamp, in
+microseconds. It is stored as a 64 bit integer in the file, and it will be
+converted to a `POSIXct` object by `read_parquet()`.
+
+To actually read the file into a data frame, and look at the first
+couple of rows, call
```
Comment (hadley) [inline content/blog/nanoparquet-0-3-0/index.Rmd:171]:
```suggestion
couple of rows, call `read_parquet()`:
```

File: content/blog/nanoparquet-0-3-0/index.Rmd [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+cc[["encodings"]][1:3]
+```
+
+## Limitations
+
+nanoparquet 0.3.0 has a number of limitations.
+
+* **Only flat tables**. `read_parquet()` can only read flat tables, i.e.
+  Parquet files where all leaf columns are children of a single root
+  column. Similarly, `write_parquet()` will not write list columns.
```
Comment (hadley) [inline content/blog/nanoparquet-0-3-0/index.Rmd:240]: Can you reframe "all leaf columns are children of a single root column" into something more informal that an R user is more likely to understand?

File: content/blog/nanoparquet-0-3-0/index.Rmd [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+  optimized, nanoparquet is probably not suited well for large data sets.
+  It should be fine for a couple of gigabytes. It may be fine if all the
+  data fits into memory comfortably.
+
+* **Single row group**. `write_parquet()` always creates a single row
+  group, which is not optimal for large files.
+
+* **Automatic encoding**. It is currently not possible to choose encodings
+  in `write_parquet()` manually.
+
```
Comment (hadley) [inline content/blog/nanoparquet-0-3-0/index.Rmd:271]: I think it's worth a note here that we'll try to tackle (some) of these over time, while continuing to keep nanoparquet lightwight. And encourage folks to submit issues if there are features that we don't support that are important for them.

**Review by hadley — APPROVED** [review]

**Comment by gaborcsardi** [comment]
Thanks! The Windows R-release binary is still not built, I'll wait a bit, hopefully they'll build it tomorrow.

**Pull request merged**